libraries/ms-common/src/lights_signal_fsm.c

```diff
@@ -80,17 +104,43 @@ static void prv_state_hazard_signal_output(Fsm *fsm, const Event *e, void *conte
   blink_event_generator_start(&storage->blink_event_generator, storage->signal_hazard_output_event);
 }
 
+static void prv_blink_event_raised_callback(BlinkerState new_state, void *context) {
+  SignalFsmStorage *storage = context;
+
+  // we only send sync events on rising edges
+  if (new_state == BLINKER_STATE_ON) {
+    storage->blink_counter++;
+
+    if (storage->sync_behaviour == SYNC_BEHAVIOUR_SEND_SYNC_MSGS &&
+        storage->blink_counter >= storage->num_blinks_between_syncs) {
+      // transmit sync event
+      storage->blink_counter = 0;
+      CAN_TRANSMIT_LIGHTS_SYNC();
+    }
+  }
+}
+
 StatusCode lights_signal_fsm_init(SignalFsmStorage *storage, const SignalFsmSettings *settings) {
+  if (settings->sync_behaviour >= NUM_SYNC_BEHAVIOURS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
   storage->signal_left_input_event = settings->signal_left_input_event;
   storage->signal_right_input_event = settings->signal_right_input_event;
   storage->signal_hazard_input_event = settings->signal_hazard_input_event;
   storage->signal_left_output_event = settings->signal_left_output_event;
   storage->signal_right_output_event = settings->signal_right_output_event;
   storage->signal_hazard_output_event = settings->signal_hazard_output_event;
+  storage->sync_event = settings->sync_event;
+  storage->sync_behaviour = settings->sync_behaviour;
+  storage->num_blinks_between_syncs = settings->num_blinks_between_syncs;
+  storage->blink_counter = 0;
 
   BlinkEventGeneratorSettings blinker_settings = {
     .interval_us = settings->blink_interval_us,
     .default_state = BLINKER_STATE_OFF,  // all lights default to off
+    .callback = &prv_blink_event_raised_callback,
```

Comment:
in C the `&` is optional for function pointers and in most of our other code we don't include it, so if you want to be a hero you could remove them from the function pointers in this file, but I won't hold you to it